src/Agent.Worker/TaskRestrictionsChecker.cs

```diff
@@ -0,0 +1,143 @@
+    {
+        public bool CheckCommand(IExecutionContext context, IWorkerCommand workerCommand, Command command)
+        {
+            ArgUtil.NotNull(context, nameof(context));
+            ArgUtil.NotNull(workerCommand, nameof(workerCommand));
+            ArgUtil.NotNull(command, nameof(command));
+
+            return Check(
+                context,
+                restrictions => !restrictions.IsCommandAllowed(workerCommand),
+                () => context.Warning(StringUtil.Loc("CommandNotAllowed", command.Area, command.Event)),
+                restrictions => PublishCommandTelemetry(context, restrictions, command));
+        }
+
+        public bool CheckSettableVariable(IExecutionContext context, string variable)
+        {
+            ArgUtil.NotNull(context, nameof(context));
+            ArgUtil.NotNull(variable, nameof(variable));
+
+            return Check(
+                context,
+                restrictions => !restrictions.IsSetVariableAllowed(variable),
+                () => context.Warning(StringUtil.Loc("SetVariableNotAllowed", variable)),
+                restrictions => PublishVariableTelemetry(context, restrictions, variable));
+        }
+
+        private bool Check(
+            IExecutionContext context,
+            Func<TaskRestrictions, bool> predicate,
+            Action warn,
+            Action<TaskDefinitionRestrictions> publishTelemetry)
+        {
+            ArgUtil.NotNull(context, nameof(context));
+            ArgUtil.NotNull(predicate, nameof(predicate));
+            ArgUtil.NotNull(warn, nameof(warn));
+            ArgUtil.NotNull(publishTelemetry, nameof(publishTelemetry));
+
+            var matches = context.Restrictions?.Where(predicate);
+
+            if (matches.IsNullOrEmpty())
+            {
+                return true;
+            }
+            else
+            {
+                var taskMatches = matches.Where(restrictions => restrictions is TaskDefinitionRestrictions).Cast<TaskDefinitionRestrictions>();
+
+                if(AgentKnobs.EnableTaskRestrictionsTelemetry.GetValue(context).AsBoolean())
+                {
+                    foreach(var match in taskMatches)
+                    {
+                        publishTelemetry(match);
+                    }
+                }
+
+                string mode = AgentKnobs.TaskRestrictionsEnforcementMode.GetValue(context).AsString();
+
+                if (String.Equals(mode, "Enabled", StringComparison.OrdinalIgnoreCase) || taskMatches.Count() != matches.Count())
+                {
+                    // we are enforcing restrictions from tasks, or we matched restrictions from the pipeline, which we always enforce
+                    warn();
+                    return false;
+                }
+                else
+                {
+                    if (!String.Equals(mode, "Disabled", StringComparison.OrdinalIgnoreCase))
+                    {
+                        warn();
```

Comment:
I think we should distinguish this warning message from the warning when the behavior is actually disabled. E.g. say something like this behavior is currently not fully enabled, but will be in April of 2021

Reply:
Good point. I'll add a different message.

Reply:
I've updated the code to differentiate the warnings.